File: src/tools/analyze_media.ts
import { z } from "zod";
import { analyzeMedia, DEFAULT_VISION_MODEL, DEFAULT_VISION_THINKING_LEVEL } from "../services/gemini_client.js";
import { isAbnormalEmpty, buildEmptyResponseWarnings, type ResponseDiagnostics } from "../utils/diagnostics.js";
import type { ToolResult } from "./registry.js";
import { getUploadStore } from "../server/upload.js";
import {
  booleanLike,
  pinnedDefaultDescription,
  pinnedThinkingDescription,
  resolveServiceTier,
  serviceTierSchema,
} from "../schemas/helpers.js";

export const analyzeMediaSchema = z.object({
  prompt: z.string().min(1),
  uploaded_file_id: z.string().optional(),
  file_uri: z.string().optional(),
  file_path: z.string().optional(),
  image_url: z.string().optional(),
  image_base64: z.string().optional(),
  model: z
    .string()
    .optional()
    .default(DEFAULT_VISION_MODEL)
    .describe(pinnedDefaultDescription("gemini_analyze_media", DEFAULT_VISION_MODEL)),
  thinking_level: z
    .enum(["minimal", "low", "medium", "high"])
    .optional()
    .default(DEFAULT_VISION_THINKING_LEVEL)
    .describe(pinnedThinkingDescription("gemini_analyze_media", DEFAULT_VISION_THINKING_LEVEL)),
  wait_for_processing: booleanLike.optional().default(true),
  service_tier: serviceTierSchema,
}).strict();

export type AnalyzeMediaArgs = z.infer<typeof analyzeMediaSchema>;

/** spec 020: Shared conversion that attaches warnings when the response is abnormally empty. */
function toResult(result: { text: string; diagnostics?: ResponseDiagnostics }): string | ToolResult {
  const { text, diagnostics } = result;
  if (diagnostics && isAbnormalEmpty(text, diagnostics)) {
    return { text, structured: { text, warnings: buildEmptyResponseWarnings(diagnostics) } };
  }
  return text;
}

export async function handleAnalyzeMedia(args: AnalyzeMediaArgs): Promise<string | ToolResult> {
  // Resolve uploaded_file_id (highest priority)
  if (args.uploaded_file_id) {
    const store = getUploadStore();
    if (!store) {
      throw new Error("UploadStore is not initialized (HTTP transport required)");
    }
    const entry = store.get(args.uploaded_file_id);
    if (!entry) {
      throw new Error(`Uploaded file not found or expired: ${args.uploaded_file_id}`);
    }
    return toResult(await analyzeMedia(args.prompt, {
      filePath: entry.storagePath,
      model: args.model,
      thinkingLevel: args.thinking_level,
      serviceTier: resolveServiceTier(args.service_tier),
    }));
  }

  if (!args.file_uri && !args.file_path && !args.image_url && !args.image_base64) {
    throw new Error("Either uploaded_file_id, file_uri, file_path, image_url, or image_base64 must be provided");
  }
  return toResult(await analyzeMedia(args.prompt, {
    fileUri: args.file_uri,
    filePath: args.file_path,
    imageUrl: args.image_url,
    imageBase64: args.image_base64,
    model: args.model,
    thinkingLevel: args.thinking_level,
    serviceTier: resolveServiceTier(args.service_tier),
  }));
}

File: src/tools/analyze_media.rs
//! `gemini_analyze_media` — multimodal analysis of images / PDF / video / audio.

use serde::Deserialize;
use serde_json::json;

use crate::schemas::{
    de_bool_like, default_true, resolve_service_tier, ServiceTierValue, ThinkingLevel,
};
use crate::services::gemini_client::{
    default_vision_model, default_vision_thinking_level, AnalyzeMediaOptions, GeminiClient,
    ThinkingSetting,
};
use crate::tools::{ToolFailure, ToolResponse};
use crate::utils::diagnostics::{build_empty_response_warnings, is_abnormal_empty};

#[derive(Debug, Deserialize, schemars::JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct AnalyzeMediaParams {
    pub prompt: String,

    #[serde(default)]
    pub file_uri: Option<String>,
    #[serde(default)]
    pub file_path: Option<String>,
    #[serde(default)]
    pub image_url: Option<String>,
    #[serde(default)]
    pub image_base64: Option<String>,

    #[serde(default = "default_vision_model")]
    #[schemars(description = "[DEFAULT FIXED] gemini_analyze_media is optimized for its purpose to run on gemini-flash-lite-latest. Do not override unless there is a clear reason (e.g. cost or a specific task's quality requirement).")]
    pub model: String,

    #[serde(default = "default_vision_thinking_level")]
    #[schemars(description = "[DEFAULT FIXED] The thinking depth of gemini_analyze_media is optimized at medium. Do not override unless there is a clear reason. Values: minimal/low/medium/high.")]
    pub thinking_level: ThinkingLevel,

    #[serde(default = "default_true", deserialize_with = "de_bool_like")]
    pub wait_for_processing: bool,

    #[serde(default)]
    pub service_tier: Option<ServiceTierValue>,
}

impl AnalyzeMediaParams {
    #[cfg(test)]
    pub fn parse(value: serde_json::Value) -> Result<Self, String> {
        let params: Self = serde_json::from_value(value).map_err(|e| e.to_string())?;
        params.validate()?;
        Ok(params)
    }

    pub fn validate(&self) -> Result<(), String> {
        if self.prompt.is_empty() {
            return Err("prompt must not be empty".to_string());
        }
        Ok(())
    }
}

pub async fn handle_analyze_media(
    client: &GeminiClient,
    params: AnalyzeMediaParams,
) -> Result<ToolResponse, ToolFailure> {
    params.validate().map_err(ToolFailure::Message)?;

    if params.file_uri.is_none()
        && params.file_path.is_none()
        && params.image_url.is_none()
        && params.image_base64.is_none()
    {
        return Err(ToolFailure::Message(
            "Either file_uri, file_path, image_url, or image_base64 must be provided".to_string(),
        ));
    }

    // Files API base64 inline path never enters a PROCESSING state, so this flag is a
    // no-op here (kept for schema compatibility with the original tool).
    let _ = params.wait_for_processing;

    let outcome = client
        .analyze_media(
            &params.prompt,
            AnalyzeMediaOptions {
                file_uri: params.file_uri.clone(),
                file_path: params.file_path.clone(),
                image_url: params.image_url.clone(),
                image_base64: params.image_base64.clone(),
                model: Some(params.model.clone()),
                service_tier: resolve_service_tier(params.service_tier),
                thinking: ThinkingSetting::Level(params.thinking_level),
            },
        )
        .await?;

    if is_abnormal_empty(&outcome.text, &outcome.diagnostics) {
        let structured = json!({
            "text": outcome.text,
            "warnings": build_empty_response_warnings(&outcome.diagnostics),
        });
        Ok(ToolResponse::with_structured(outcome.text, structured))
    } else {
        Ok(ToolResponse::text(outcome.text))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn applies_pinned_defaults() {
        let p = AnalyzeMediaParams::parse(json!({ "prompt": "x", "file_path": "/tmp/x.png" })).unwrap();
        assert_eq!(p.model, "gemini-flash-lite-latest");
        assert_eq!(p.thinking_level, ThinkingLevel::Medium);
        assert!(p.wait_for_processing);
    }

    #[test]
    fn descriptions_have_default_fixed_marker() {
        let schema = serde_json::to_value(schemars::schema_for!(AnalyzeMediaParams)).unwrap();
        for field in ["model", "thinking_level"] {
            assert!(schema["properties"][field]["description"]
                .as_str()
                .unwrap_or_default()
                .contains("[DEFAULT FIXED]"));
        }
    }
}
